src/backend/PortabilityService.ConfigurationSource/ConfigurationSource.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Extensions.Configuration.Memory;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace PortabilityService.ConfigurationSource
+{
+    public class PortabilityServiceConfigurationSource : MemoryConfigurationSource
+    {
+        private const string DefaultConfigurationServiceUrl = "http://localhost:1443";
+        internal const string PortabilityServiceConfigurationRoot = "PortabilityServiceSettings";
+
+        public PortabilityServiceConfigurationSource(string configurationSection = PortabilityServiceConfigurationRoot)
+            : this(GetConfigurationServiceUrlFromEnvironment() ?? new Uri(DefaultConfigurationServiceUrl), configurationSection) { }
+
+        public PortabilityServiceConfigurationSource(Uri configurationServiceUrl, string configurationSection = PortabilityServiceConfigurationRoot)
+            : this(CreateHttpClient(configurationServiceUrl ?? throw new ArgumentNullException(nameof(configurationServiceUrl))), configurationSection ?? throw new ArgumentNullException(nameof(configurationSection))) { }
+
+        public PortabilityServiceConfigurationSource(HttpClient httpClient, string configurationSection = PortabilityServiceConfigurationRoot)
+        {
+            InitialData = GetConfigurationSectionSettings(configurationSection ?? throw new ArgumentNullException(nameof(configurationSection)), httpClient ?? throw new ArgumentNullException(nameof(httpClient))).GetAwaiter().GetResult();
```

Comment:
nit: Separate into new lines to make this easier to follow:

```csharp
InitialData = GetConfigurationSectionSettings(
    configurationSection ?? throw new ArgumentNullException(nameof(configurationSection)),
    httpClient ?? throw new ArgumentNullException(nameof(httpClient))
)
.GetAwaiter().GetResult();
```

I would replace the initialization with an AsyncLazy<T>... Typically, you want your constructors to return quickly and async methods can block for an indeterminate period of time while you're waiting for the result.
https://blog.stephencleary.com/2013/01/async-oop-2-constructors.html

Reply:
Agreed on not waiting for initialization in the ctor. Unfortunately, I don't know a good way to make ASP.NET Core configuration providers/sources async. One option might be to just store the URL and section name here and wait on retrieving the settings in a custom configuration provider's `Load` method. It's still not async, but at least the synchronous blocking is not in the ctor. This is what first-party providers like the [KeyVault config source/provider](https://github.com/aspnet/Configuration/tree/dev/src/Config.AzureKeyVault) do.